### PR TITLE
[kube-prometheus-stack] Update prometheus-operator to 0.42.1

### PR DESCRIPTION
**Pull request by bashofmann**
#### What this PR does / why we need it:
Update prometheus-operator to 0.42.1.

This also introduces a new `Probes` CRD. Since Helm does not create this new CRD on a chart upgrade automatically, the chart version is bumped to 10 to indicate the breaking change.

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Comment by cablespaghetti** [comment]
Would love for this to get merged. I've done this by hand by bumping image versions and deploying the CRDs by hand but the lack of "probes" in the ClusterRole makes things difficult.

**Comment by tiagomeireles** [comment]
This also bumps the kubernetes prerequisite to 1.16+. prometheus-operator 0.39 introduced a requirement for kubernetes v1.16 or newer. https://github.com/prometheus-operator/prometheus-operator/releases/tag/v0.39.0

I'm +1 with getting this merged in.

**Comment by bashofmann** [comment]
What do you think, should I add a `kubeVersion` field in the Chart.yaml to restrict this?

**Comment by bashofmann** [comment]
`kubeVersion` requirement is added.

**Comment by scottrigby** [comment]
Related to https://github.com/prometheus-community/helm-charts/issues/24

**Comment by scottrigby** [comment]
Have these CRDs been updated in the upstream https://github.com/prometheus-operator/kube-prometheus project? This may be a question for @brancz

**Comment by bashofmann** [comment]
I took the CRDs in this PR directly from here: https://github.com/prometheus-operator/prometheus-operator/tree/v0.42.1/example/prometheus-operator-crd

**Comment by scottrigby** [comment]
@bashofmann OK thanks yes, it appears the source – generated as part of the `prometheus-operator/kube-prometheus` project (which this chart installs) – is here: https://github.com/prometheus-operator/kube-prometheus/tree/master/manifests/setup

These were updated recently upstream:

- to `0.42.0` https://github.com/prometheus-operator/kube-prometheus/pull/674
- to `0.42.1` https://github.com/prometheus-operator/kube-prometheus/pull/685

Which is interesting to me. I wonder:
1. why the chart [README](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/README.md#helm-fails-to-create-crds) recommends installing directly from `https://raw.githubusercontent.com/prometheus-operator/prometheus-operator` as opposed to from the `kube-prometheus` generated manifests
1. and why the chart [crds](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack/crds) dir files have [similar comments](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/crds/crd-alertmanager.yaml#L1) listing `prometheus-operator/prometheus-operator` as the file source
1. Also, this chart's `hack` dir has sync scripts for grafana dashboards and prom rules, but not for CRDs – should it? 

Maybe some of the other `kube-prometheus` project maintainers can help me understand if these are just relics/gaps left over from the stable chart, or if there is some reason. Either way I think we can clean this up and decide on a nicer path forward to keep better in sync with the upstream `kube-prometheus` project. CC @metalmatze @paulfantom @lilic

**Comment by bashofmann** [comment]
Are you sure that `kube-prometheus` is the source of truth for the CRDs? It feels more logical to me that the source of truth is in `prometheus-operator` which uses and implements logic for these CRDs and not in `kube-prometheus` which bundles the operator with monitors, dashboards and rules.

**Comment by paulfantom** [comment]
The source of truth for CRDs is their [jsonnet definition in prometheus-operator repository](https://github.com/prometheus-operator/prometheus-operator/tree/master/jsonnet/prometheus-operator), this is used to generate [yaml files in prometheus-operator](https://github.com/prometheus-operator/prometheus-operator/tree/master/example/prometheus-operator-crd) and it is also consumed by kube-prometheus.

**Comment by Nuru** [comment]
This chart fails on AWS EKS and probably fails on GKE due to the `kubeVersion` constraint:
```
Error: UPGRADE FAILED: chart requires kubeVersion: >=1.16.0 which is incompatible with Kubernetes v1.17.9-eks-4c6976
```
Solution appears to be to change the constraint to `>=1.16.0-0`. See https://github.com/helm/helm/issues/3810#issuecomment-379877753 and [this documentation](https://github.com/Masterminds/semver#checking-version-constraints)

**Comment by bashofmann** [comment]
Thanks, good catch. I fixed the constraint and rebased the PR.

**Comment by deleted** [comment]
I would also love to see this PR land, as `prometheus-operator` v0.39 added support to enable the prometheus query logger.

**Comment by uptickmetachu** [comment]
(Also keenly interested in this one for black box exporter)

**Comment by bashofmann** [comment]
@scottrigby Is there anything else this PR needs besides the obvious rebase?

File: charts/kube-prometheus-stack/values.yaml [hunk 1]
```diff
@@ -1625,7 +1629,7 @@ prometheus:
     ##
     image:
       repository: quay.io/prometheus/prometheus
-      tag: v2.18.2
+      tag: v2.19.2
```
Comment (monotek) [inline charts/kube-prometheus-stack/values.yaml:1632]: this could be v2.21.0

**Review by monotek — CHANGES_REQUESTED** [review]
resolve merge conflict please

File: charts/kube-prometheus-stack/values.yaml (hunk 1 above)
Comment (bashofmann) [reply charts/kube-prometheus-stack/values.yaml]: Done

**Review by monotek — DISMISSED** [review]

**Comment by bashofmann** [comment]
I rebased the branch again to resolve conflicts.

**Comment by bashofmann** [comment]
The lint task seems to have failed due to a network hickup. Is there a way to rerun it (with the right permissions), or do I need to make a tiny change and push again?

**Review by monotek — CHANGES_REQUESTED** [review]
@bashofmann 

Could you merge the changes from master to your branch introduced by #191 please?
After that linting should succeed :)

**Comment by bashofmann** [comment]
@monotek rebased and pushed

**Review by monotek — APPROVED** [review]

**Review by scottrigby — APPROVED** [review]
Thanks for your work on this

**Pull request merged**